### PR TITLE
Webpage extension patch

### PR DESCRIPTION
**Pull request by LarsPauli**
Added timer capabilities.

### CODE REVIEW

File: main/esp32_ssd1306.c [hunk 1]
```diff
@@ -123,6 +127,8 @@ static esp_err_t ssd1306_write_single_command(uint8_t command)
 static esp_err_t ssd1306_write_command(uint8_t command)
 {
     if (ssd1306_dev.i2c_cmd) {
+        ESP_ERROR_CHECK(i2c_master_write_byte(ssd1306_dev.i2c_cmd,
+                                              0b10000000, true));
```
Comment (alxhoff) [inline main/esp32_ssd1306.c:131]: again here

**Review by alxhoff — COMMENTED** [review]
I am going to fix up a few of the small things and push them to the PR first

**Comment by alxhoff** [comment]
I tried to push my changes but I believe you need to do this

https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

File: main/paf_gpio.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
+/*
```
[the remark this answers is not in the document]
Comment (LarsPauli) [reply main/paf_gpio.c:1]: We could use the mode enum for changing this behavior, I'm going to look into that later.

File: main/esp32_ssd1306.c [hunk 3]
```diff
@@ -379,7 +386,38 @@ signed char ssd1306_init(void)
     ssd1306_dev.background = SSD1306_BACKGROUND;
     ssd1306_dev.font = SSD1306_FONT;
 
-    /* Init LCD */
+    /*
```
[the remark this answers is not in the document]
Comment (LarsPauli) [reply main/esp32_ssd1306.c:389]: Some random stuff I overlooked during formating the files

File: main/esp32_ssd1306.c [hunk 4]
```diff
@@ -103,6 +104,7 @@ static esp_err_t ssd1306_write_start(void)
 static esp_err_t ssd1306_write_end(void)
 {
     esp_err_t ret = i2c_master_stop(ssd1306_dev.i2c_cmd);
+    ret = i2c_master_cmd_begin(PAF_DEF_I2C_NUM,ssd1306_dev.i2c_cmd,100);
```
[the remark this answers is not in the document]
Comment (LarsPauli) [reply main/esp32_ssd1306.c:107]: The I2C hardware is not outputting the values in the buffers untill i2c_master_cmd_begin is called. So it is needed for the function of the bus.

File: main/esp32_ssd1306.c [hunk 5]
```diff
@@ -112,6 +114,8 @@ static esp_err_t ssd1306_write_single_command(uint8_t command)
 {
     if (ssd1306_dev.i2c_cmd) {
         ssd1306_write_start();
+        ESP_ERROR_CHECK(i2c_master_write_byte(ssd1306_dev.i2c_cmd,
+                                              0b01000000, true));
```
[the remark this answers is not in the document]
Comment (LarsPauli) [reply main/esp32_ssd1306.c:118]: The control byte as defined in the datasheet of the ssd1306. Its needed in front of each command and data byte to show the type of data transmitted in the following byte (data or cmd). 
Sorry for using magic numbers.

File: main/paf_led.c [hunk 6]
```diff
@@ -62,6 +67,38 @@ struct led_config {
                       .ledc_freq = PAF_DEF_LED_FREQ
                     };
 
+struct pulsGen_config{
+    timer_config_t hw_timer1_config;
+    unsigned int pulse_on_duraton; 
+    unsigned int pulse_off_duration;
+    unsigned int pulse_periode;
+} pulsGen_cfg.hw_timer1_config = {
```
[the remark this answers is not in the document]
Comment (LarsPauli) [reply main/paf_led.c]: This is fixed in the current push.

File: main/paf_led.c [hunk 7]
```diff
@@ -273,3 +316,66 @@ unsigned int paf_led_get_freq(void)
...
+
+esp_err_t paf_led_init_pulse()
+{
+    // check if T_PWM < T_pulse 
+
+
+    //setup timer 0
+    timer_init(TIMER_GROUP_1,TIMER_0, &pulsGen_cfg.hw_timer1_config);
+    timer_set_counter_value(TIMER_GROUP_1,TIMER_0, 0);
+    timer_set_alarm_value(TIMER_GROUP_1,TIMER_0, pulsGen_cfg.pulse_duration_off);
```
[the remark this answers is not in the document]
Comment (LarsPauli) [reply main/paf_led.c]: allready fixed in the current push

File: main/esp32_ssd1306.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (alxhoff) [reply main/esp32_ssd1306.c:107]: ahhh I must of missed that when doing my port

File: main/esp32_ssd1306.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (alxhoff) [reply main/esp32_ssd1306.c:118]: ahhh makes sense

**Comment by alxhoff** [comment]
I haven't tested this yet on hardware but I will go through the I2C command stuff tonight and let you know when it's done to test. Also rebased and stripped whitespace. You can also format you code with the format target i documented in the README.

**Comment by alxhoff** [comment]
> I tried to push my changes but I believe you need to do this
> 
> https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

also forgot I changed my GH password ops

**Pull request merged**